### PR TITLE
Check for ClearValue without LoadOpClear

### PR DESCRIPTION
**Pull request by paul-lunarg**
Add warning message when VkRenderPassBeginInfo.clearValueCount>0
but none of the attachments have VK_ATTACHMENT_LOAD_OP_CLEAR
Close https://github.com/KhronosGroup/Vulkan-ValidationLayers/issues/3123

### CODE REVIEW

File: layers/best_practices_utils.cpp [hunk 1]
```diff
@@ -1697,6 +1697,28 @@ bool BestPractices::ValidateCmdBeginRenderPass(VkCommandBuffer commandBuffer, Re
...
+                clearing = true;
+                break;
+            }
+        }
+
+        if (!clearing && pRenderPassBegin->clearValueCount > 0) {
+            // Isn't clearing but has clearValues
+            skip |= LogWarning(
+                device, kVUID_BestPractices_ClearValueWithoutLoadOpClear,
+                "This render pass does not have VkRenderPassCreateInfo.pAttachments->loadOp == VK_ATTACHMENT_LOAD_OP_CLEAR "
```
Comment (ncesario-lunarg) [inline layers/best_practices_utils.cpp:1718]: Would it be worth spelling out a bit more? e.g., "...does not have _any_ VkRenderPassCreateInfo.pAttachments with loadOp == VK_ATTACHMENT_LOAD_OP_CLEAR.".

File: tests/vklayertests_best_practices.cpp [hunk 2]
```diff
@@ -1572,3 +1572,97 @@ TEST_F(VkBestPracticesLayerTest, OverAllocateFromDescriptorPool) {
...
     m_errorMonitor->VerifyFound();
 }
+
+TEST_F(VkBestPracticesLayerTest, RenderPassClearWithoutLoadOpClear) {
+    TEST_DESCRIPTION("Test for clearing a RenderPass with non-zero clearValueCount without any VK_ATTACHMENT_LOAD_OP_CLEAR");
+
+    InitBestPracticesFramework();
+    InitState();
+
+    m_errorMonitor->ExpectSuccess();
```
Comment (ncesario-lunarg) [inline tests/vklayertests_best_practices.cpp:1582]: Might as well put this right at the top, right after `TEST_DESCRIPTION` to catch _any_ errors (as unlikely as they might be).

File: tests/vklayertests_best_practices.cpp [hunk 3]
```diff
@@ -1572,3 +1572,97 @@ TEST_F(VkBestPracticesLayerTest, OverAllocateFromDescriptorPool) {
...
+
+    VkImageViewCreateInfo iv_info = LvlInitStruct<VkImageViewCreateInfo>();
+    iv_info.format = VK_FORMAT_R8G8B8A8_UNORM;
+    iv_info.image = image.handle();
+    iv_info.viewType = VK_IMAGE_VIEW_TYPE_2D;
+    iv_info.subresourceRange = {VK_IMAGE_ASPECT_COLOR_BIT, 0, 1, 0, 1};
+    iv_info.components = {VK_COMPONENT_SWIZZLE_R, VK_COMPONENT_SWIZZLE_G, VK_COMPONENT_SWIZZLE_B, VK_COMPONENT_SWIZZLE_A};
+
+    VkImageView image_view = VK_NULL_HANDLE;
+    vk::CreateImageView(m_device->device(), &iv_info, nullptr, &image_view);
```
Comment (ncesario-lunarg) [inline tests/vklayertests_best_practices.cpp:1608]: I think you can make use of `VkImageObj::targetView` here. e.g., `const auto image_view = image.targetView(VK_FORMAT_R8G8B8A8_UNORM);`

File: tests/vklayertests_best_practices.cpp [hunk 4]
```diff
@@ -1572,3 +1572,97 @@ TEST_F(VkBestPracticesLayerTest, OverAllocateFromDescriptorPool) {
...
+    spd.pColorAttachments = &ar;
+
+    VkRenderPassCreateInfo rp_info = LvlInitStruct<VkRenderPassCreateInfo>();
+    rp_info.attachmentCount = 1;
+    rp_info.pAttachments = &attachment;
+    rp_info.subpassCount = 1;
+    rp_info.pSubpasses = &spd;
+
+    VkRenderPass rp;
+    vk::CreateRenderPass(m_device->device(), &rp_info, nullptr, &rp);
```
Comment (ncesario-lunarg) [inline tests/vklayertests_best_practices.cpp:1634]: Where possible, best to use the test wrappers, e.g., `vk_testing::RenderPass rp(*m_device, rp_info);` here. You'll just need to use the `handle()` function to access the object's "raw" Vulkan handle. This helps to ensure these objects get cleaned up properly (I don't think this is currently getting destroyed).

File: tests/vklayertests_best_practices.cpp [hunk 5]
```diff
@@ -1572,3 +1572,97 @@ TEST_F(VkBestPracticesLayerTest, OverAllocateFromDescriptorPool) {
...
+    VkFramebufferCreateInfo fb_info = LvlInitStruct<VkFramebufferCreateInfo>();
+    fb_info.width = w;
+    fb_info.height = h;
+    fb_info.layers = 1;
+    fb_info.renderPass = rp;
+    fb_info.attachmentCount = 1;
+    fb_info.pAttachments = &image_view;
+
+    VkFramebuffer fb;
+    vk::CreateFramebuffer(m_device->device(), &fb_info, NULL, &fb);
```
Comment (ncesario-lunarg) [inline tests/vklayertests_best_practices.cpp:1645]: Same comment as above; you can use `vk_testing::Framebuffer (*m_device, fb_info);` to ensure proper cleanup.

File: tests/vklayertests_best_practices.cpp [hunk 6]
```diff
@@ -1572,3 +1572,97 @@ TEST_F(VkBestPracticesLayerTest, OverAllocateFromDescriptorPool) {
...
+
+    m_commandBuffer->begin();
+
+    VkClearValue cv{};
+    cv.color = VkClearColorValue{};
+    memset(cv.color.float32, 0, 4 * 4);
```
Comment (ncesario-lunarg) [inline tests/vklayertests_best_practices.cpp:1651]: I'd vote for something with some type awareness such as `std::fill` over `::memset`.

**Review by ncesario-lunarg — CHANGES_REQUESTED** [review]
Great to see this PR up so fast @paul-lunarg! Just some minor nits with regards to the test, but otherwise this looks good!

File: tests/vklayertests_best_practices.cpp (hunk 2 above)
Comment (paul-lunarg) [reply tests/vklayertests_best_practices.cpp:1582]: Sure, didn't know if I could, I'll add asserts on the Init calls too

File: layers/best_practices_error_enums.h [hunk 7]
```diff
@@ -30,6 +30,8 @@
 #define DECORATE_UNUSED
 #endif
 
+// clang-format off
```
[the remark this answers is not in the document]
Comment (paul-lunarg) [reply layers/best_practices_error_enums.h:33]: I added clang-format on but didn't change to pragma once, it should probably be in a different PR, and discussed with others

File: layers/best_practices_utils.cpp (hunk 1 above)
Comment (ncesario-lunarg) [reply layers/best_practices_utils.cpp:1719]: `VkRenderPassCreateInfo.pAttachments->loadOp` is good with me if it's also clear to others; just a bit different from what I've seen before.

File: tests/vklayertests_best_practices.cpp [hunk 8]
```diff
@@ -1572,3 +1572,99 @@ TEST_F(VkBestPracticesLayerTest, OverAllocateFromDescriptorPool) {
...
+
+    VkRenderPassBeginInfo begin_info = LvlInitStruct<VkRenderPassBeginInfo>();
+    begin_info.clearValueCount = 1;  // Pass one clearValue, in conflict with attachment.loadOp = VK_ATTACHMENT_LOAD_OP_DONT_CARE
+    begin_info.pClearValues = &cv;
+    begin_info.renderPass = rp.handle();
+    begin_info.renderArea.extent.width = w;
+    begin_info.renderArea.extent.height = h;
+    begin_info.framebuffer = fb.handle();
+
+    // Setup finished
```
Comment (ncesario-lunarg) [inline tests/vklayertests_best_practices.cpp:1657]: I like this! Sometimes it's difficult to tell what's actually being tested; wonder if we should make this more common practice.

**Review by ncesario-lunarg — APPROVED** [review]
LGTM! Nice work!

**Pull request merged**